### PR TITLE
[TTAHUB-27]: Frontend changes for multiple approvers

### PR DESCRIPTION
**Pull request by dcloud**
## Description of change

Frontend changes to support multiple approvers for an activity report.

**Currently pointing a draft branch for backend work**

Most of the frontend work lies in files within `frontend/src/pages/ActivityReport/Pages/Review/`, but we will also need to update various fetchers in `frontend/src/fetchers` to use the new request/reponse formats.

**DEV NOTES:**
- One or many approvers can be added to a AR.
- In the 'draft' state approvers cane be added, removed, and saved.
- Each approver can set their review and comment notes.
- If an approver is removed their comments are saved. If added back they will retain their original comments.
- Every manager who has approved remains approved regardless of report status 'needs_action', 'draft'.
- If anyone of the managers marks the report as 'needs_action' the report is set to the 'needs_action' state. In this state the creator or collaborators can change the report.
- An approver can change their review _and_ comment anytime before the report is fully approved.

## How to test

To test please pull down this branch locally. Run both db:migrate and db:seed on a clean database.

**Test the following cases:**
If you encounter an error or something you think is an error please note the **EXACT** steps to reproduce.

Draft - Saving / Updating / Deleting (data and multiple approvers)
Submitting - Approved / Pending / Re Submit / Reset to Draft
Needs Action - Submitter / Approved / Pending
Approved - Submitter / Approver

Please test any additional test cases that you feel are valid. 

## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-27

### Every PR

- [x] Meets issue criteria
- [x] JIRA ticket status updated
- [x] Code is meaningfully tested
- [x] Meets accessibility standards (WCAG 2.1 Levels A, AA)
~- [ ] API Documentation updated~
~- [ ] Boundary diagram updated~
~- [ ] Logical Data Model updated~
~- [ ] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions~

### Production Deploy

~- [ ] Staging smoke test completed~

### After merge/deploy

- [x] Update JIRA ticket status

### CODE REVIEW

File: frontend/src/pages/ActivityReport/Pages/Review/__tests__/ReviewPage.js
Comment (thewatermethod) [inline frontend/src/pages/ActivityReport/Pages/Review/__tests__/ReviewPage.js:109]: remove comment?

**Review by thewatermethod — APPROVED** [review]
Big job - nice work! I created a couple activity reports with multiple approvers and ran through the process a couple of times. I didn't notice any bugs. Approved reports display fine. Suggested a few minor things to change overall.

Might be worth having someone else review in addition, given the size of the PR.

File: frontend/src/fetchers/__tests__/activityReports.js
Comment (AdamAdHocTeam) [reply frontend/src/fetchers/__tests__/activityReports.js:93]: Updated test to check JSON.

File: frontend/src/pages/ActivityReport/Pages/Review/__tests__/ReviewPage.js
Comment (AdamAdHocTeam) [reply frontend/src/pages/ActivityReport/Pages/Review/__tests__/ReviewPage.js]: Bold tags replaced with Strong tags.

File: frontend/src/pages/ActivityReport/Pages/Review/Submitter/NeedsAction.js
Comment (AdamAdHocTeam) [reply frontend/src/pages/ActivityReport/Pages/Review/Submitter/NeedsAction.js]: Bold tags replaced with Strong tags.

File: frontend/src/pages/ActivityReport/Pages/Review/Approver/Approved.js
Comment (AdamAdHocTeam) [reply frontend/src/pages/ActivityReport/Pages/Review/Approver/Approved.js]: Bold tags replaced with Strong tags.

**Review by jasalisbury — COMMENTED** [review]
Lots of changes! Still need to look at the code but some minor things I found playing around with it so far:

- When report is submitted, for the submitter the top alert shows `(x of y)` does it need "reviews pending" after?
- When approving, selecting "needs action" makes `edit` buttons appear
- I can't remember, if a manager selects needs action now does that report show up in the managers "Alert" list?
- Suppose two pending approvers. Manager 1 approves. Submitter resets to draft. Manager 1 approval still is applied to the report. Is that expected behavior? Or do we want to reset that managers approval back to pending? I could go either way...

**Comment by AdamAdHocTeam** [comment]
> Lots of changes! Still need to look at the code but some minor things I found playing around with it so far:
> 
> * When report is submitted, for the submitter the top alert shows `(x of y)` does it need "reviews pending" after?
> * When approving, selecting "needs action" makes `edit` buttons appear
> * I can't remember, if a manager selects needs action now does that report show up in the managers "Alert" list?
> * Suppose two pending approvers. Manager 1 approves. Submitter resets to draft. Manager 1 approval still is applied to the report. Is that expected behavior? Or do we want to reset that managers approval back to pending? I could go either way...

- I have removed the "reviews pending".
- Fixed issue where edit link was incorrectly showing.
- If a report is flagged as needs_action it should show in the alert list.
- Per Patrice if a approver 'approves' the report we will keep their 'approved' state even if changes are made.

**Comment by kryswisnaskas** [comment]
Looks good 🎉 ! I like how the alerts table shows the number of approvals. I verified that multiple scenarios are supported. Nice work!

One request - 
Could we adjust the text on the summary page for the approvers? Currently, it shows the following:

_

> Karen Johns has requested approval for this activity report (2 of 3).
> Please review all information in each section before submitting for approval.

_
How about changing it to say 

> Please review all information in each section before approval.

_

File: frontend/src/components/MultiSelect.js
Comment (kryswisnaskas) [inline frontend/src/components/MultiSelect.js:110]: Curious - any reason lodash was needed here?

Comment (AdamAdHocTeam) [reply frontend/src/components/MultiSelect.js:110]: This was changed to use loddash in order to access properties on the User object of the approver (approver.User.id).

**Review by kryswisnaskas — APPROVED** [review]
👍

**Comment by jasalisbury** [comment]
> * I have removed the "reviews pending".

I should have been clearer, I think we should keep "reviews pending" for the approver and _add_ it to the submitters view, just to help clarify what the (x of y) means.

**Comment by AdamAdHocTeam** [comment]
> > * I have removed the "reviews pending".
> 
> I should have been clearer, I think we should keep "reviews pending" for the approver and _add_ it to the submitters view, just to help clarify what the (x of y) means.

I have add it back to approver and the submitter.

File: src/services/activityReports.js
Comment (jasalisbury) [inline src/services/activityReports.js:475]: This feels more like a concern of the UI then of the API.

**Review by jasalisbury — COMMENTED** [review]
Great job getting all these changes in, you had to navigate a lot of complexity. Most my comments are pretty minor.

File: src/services/activityReports.js
Comment (AdamAdHocTeam) [reply src/services/activityReports.js:475]: This was originally written on the front end but was moved to the backed end. This is because certain approvers don't have the complete list of approvers and status from the alerts sql.

File: frontend/src/components/SingleReadOnlyEditor.js
Comment (AdamAdHocTeam) [reply frontend/src/components/SingleReadOnlyEditor.js:1]: Changed name to 'ReadOnlyEditor'.

File: frontend/src/pages/ActivityReport/Pages/Review/Approver/Review.js
Comment (AdamAdHocTeam) [reply frontend/src/pages/ActivityReport/Pages/Review/Approver/Review.js:80]: Agreed, changed to use 'hasBeenReviewed'.

File: frontend/src/pages/ActivityReport/Pages/Review/Approver/Review.js
Comment (AdamAdHocTeam) [reply frontend/src/pages/ActivityReport/Pages/Review/Approver/Review.js:95]: Agreed, changed to use 'hasBeenReviewed'.

File: frontend/src/pages/ActivityReport/Pages/Review/Submitter/index.js
Comment (AdamAdHocTeam) [reply frontend/src/pages/ActivityReport/Pages/Review/Submitter/index.js:54]: Agree switched to map (much cleaner).

File: frontend/src/pages/ActivityReport/index.js
Comment (AdamAdHocTeam) [reply frontend/src/pages/ActivityReport/index.js:250]: Yes, the multi select is changing the approver object to have a properties like 'User.Id'. 

So the work around here was to use loddash. 

Let me know if you want to discuss further.

File: frontend/src/pages/ApprovedActivityReport/components/ViewTable.css
Comment (AdamAdHocTeam) [reply frontend/src/pages/ApprovedActivityReport/components/ViewTable.css:28]: Blank line removed.

File: frontend/src/pages/ActivityReport/index.js
Comment (AdamAdHocTeam) [reply frontend/src/pages/ActivityReport/index.js]: This has been fixed with the changes to multi select.

File: frontend/src/pages/ActivityReport/index.js
Comment (AdamAdHocTeam) [reply frontend/src/pages/ActivityReport/index.js]: Added code and unit test for keeping approvers collection the same.

**Review by jasalisbury — APPROVED** [review]
:ship:

**Pull request merged**
